libc/zipos/to_fd.c

```diff
@@ -0,0 +1,80 @@
+╞══════════════════════════════════════════════════════════════════════════════╡
+│ Copyright 2023 Justine Alexandra Roberts Tunney                              │
+│                                                                              │
+│ Permission to use, copy, modify, and/or distribute this software for         │
+│ any purpose with or without fee is hereby granted, provided that the         │
+│ above copyright notice and this permission notice appear in all copies.      │
+│                                                                              │
+│ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL                │
+│ WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED                │
+│ WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE             │
+│ AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL         │
+│ DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR        │
+│ PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER               │
+│ TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR             │
+│ PERFORMANCE OF THIS SOFTWARE.                                                │
+╚─────────────────────────────────────────────────────────────────────────────*/
+#include "libc/calls/calls.h"
+#include "libc/calls/struct/iovec.h"
+#include "libc/calls/struct/stat.h"
+#include "libc/calls/syscall-sysv.internal.h"
+#include "libc/dce.h"
+#include "libc/errno.h"
+#include "libc/runtime/runtime.h"
+#include "libc/sysv/consts/map.h"
+#include "libc/sysv/consts/mfd.h"
+#include "libc/sysv/consts/o.h"
+#include "libc/sysv/consts/prot.h"
+#include "libc/sysv/consts/shm.h"
+#include "libc/sysv/errfuns.h"
+#include "libc/zipos/zipos.internal.h"
+
+/**
+ * Create a file descriptor and copy a zipos file to it.
+ *
+ * This is only supported on Linux and FreeBSD (fexecve platforms).
+ *
+ * @param h is zipos file and current file position is ignored
+ * @return a fd valid to pass to fexecve on success, otherwise -1 w/ errno
+ * @raise ENOSYS on Windows, XNU, OpenBSD, NetBSD, and Metal
+ */
+int __zipos_to_fd(struct ZiposHandle *h) {
+  if (!IsLinux() && !IsFreebsd()) {
+    return enosys();
+  }
+
+  const int64_t beforeOffset = __zipos_lseek(h, 0, SEEK_CUR);
+  struct stat st;
+  if ((beforeOffset == -1) || (__zipos_fstat(h, &st) == -1)) {
+    return -1;
+  }
+  int fd;
+  if (IsLinux()) {
+    fd = memfd_create(__func__, MFD_CLOEXEC);
+  } else /*if(IsFreebsd())*/ {
+    fd = sys_shm_open(SHM_ANON, O_CREAT | O_RDWR, 0);
+  }
+  if (fd == -1) {
+    return -1;
+  }
+  void *space;
+  if ((ftruncate(fd, st.st_size) != -1) &&
+      ((space = mmap(0, st.st_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd,
+                     0)) != MAP_FAILED)) {
+    const bool success =
+        (__zipos_read(h, &(struct iovec){space, st.st_size}, 1, 0) != -1) &&
+        (__zipos_lseek(h, beforeOffset, SEEK_SET) != -1);
```

Comment:
`__zipos_lseek` shouldn't be necessary since `__zipos_read` only changes the file position if the last argument is `-1`. Since you're specifying `0` as the last arg, `__zipos_read` will have `pread()` behavior instead.

Reply:
That's handy and simplifies the `mmap` too